internal/querynode/query_shard.go

```diff
@@ -30,27 +46,816 @@ type queryShard struct {
+	segmentIDs := req.SegmentIDs
+	timestamp := req.Req.TravelTimestamp
+
+	// check ctx timeout
+	if !funcutil.CheckCtxValid(ctx) {
+		return nil, errors.New("search context timeout")
+	}
+
+	// check if collection has been released
+	collection, err := q.historical.replica.getCollectionByID(collectionID)
+	if err != nil {
+		return nil, err
+	}
+	if req.GetReq().GetGuaranteeTimestamp() >= collection.getReleaseTime() {
+		log.Warn("collection release before search", zap.Int64("collectionID", collectionID))
+		return nil, fmt.Errorf("retrieve failed, collection has been released, collectionID = %d", collectionID)
+	}
+
+	// deserialize query plan
+
+	var plan *SearchPlan
+	if req.Req.GetDslType() == commonpb.DslType_BoolExprV1 {
+		expr := req.Req.SerializedExprPlan
+		plan, err = createSearchPlanByExpr(collection, expr)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		dsl := req.Req.Dsl
+		plan, err = createSearchPlan(collection, dsl)
+		if err != nil {
+			return nil, err
+		}
+	}
+	defer plan.delete()
+
+	schemaHelper, err := typeutil.CreateSchemaHelper(collection.schema)
+	if err != nil {
+		return nil, err
+	}
+
+	// validate top-k
+	topK := plan.getTopK()
+	if topK <= 0 || topK >= 16385 {
+		return nil, fmt.Errorf("limit should be in range [1, 16385], but got %d", topK)
+	}
+
+	// parse plan to search request
+	searchReq, err := parseSearchRequest(plan, req.Req.PlaceholderGroup)
+	if err != nil {
+		return nil, err
+	}
+	defer searchReq.delete()
+	queryNum := searchReq.getNumOfQuery()
+	searchRequests := []*searchRequest{searchReq}
+
+	if len(segmentIDs) == 0 {
+		// segmentIDs not specified, searching as shard leader
+		return q.searchLeader(ctx, req, searchRequests, collectionID, schemaHelper, plan, topK, queryNum, timestamp)
+	}
+
+	// segmentIDs specified search as shard follower
+	return q.searchFollower(ctx, req, searchRequests, collectionID, schemaHelper, plan, topK, queryNum, timestamp)
+}
+
+func (q *queryShard) searchLeader(ctx context.Context, req *querypb.SearchRequest, searchRequests []*searchRequest, collectionID UniqueID,
+	schemaHelper *typeutil.SchemaHelper, plan *SearchPlan, topK int64, queryNum int64, timestamp Timestamp) (*internalpb.SearchResults, error) {
+	q.streaming.replica.queryRLock()
+	defer q.streaming.replica.queryRUnlock()
+	cluster, ok := q.clusterService.getShardCluster(req.GetDmlChannel())
+	if !ok {
+		return nil, fmt.Errorf("channel %s leader is not here", req.GetDmlChannel())
+	}
+
+	// shard leader dispatches request to its shard cluster
+	results, err := cluster.Search(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	// hold request until guarantee timestamp >= service timestamp
+	guaranteeTs := req.GetReq().GetGuaranteeTimestamp()
+	q.waitUntilServiceable(ctx, guaranteeTs, tsTypeDML)
+	// shard leader queries its own streaming data
+	streamingResults, _, _, err := q.streaming.search(searchRequests, collectionID, req.Req.PartitionIDs, req.DmlChannel, plan, timestamp)
```

Comment:
streaming search should happen together with cluster.search.
It doesn't make sense to wait for cluster.search done before start streaming search

Reply:
Fixed